README.md


# arm_now 1.2
arm_now is a qemu powered tool that allows instant setup of virtual machines on arm cpu, mips, powerpc, nios2, x86 and more, for reverse, exploit, fuzzing and programming purpose.

![Alt Text](https://github.com/nongiach/arm_now/blob/assets/arm_now.gif)

# Install
```sh
# pip3 install https://github.com/nongiach/arm_now/archive/master.zip --upgrade
# # Or you can also do:
# pip3 install arm_now
```
Supported System: Linux, Windows WSL, MacOS, Docker.
# Docker install
If you are using docker, just run and enjoy!
```sh
$ docker run -it --name arm_now -v $PWD:/mount --rm bannsec/arm_now_docker arm_now
```
For a bash function wrapper see here: https://github.com/bannsec/arm_now_docker 

# Start an arm Virtual Machine
```sh
$ arm_now start armv5-eabi
Welcome to arm_now
buildroot login: root
# uname -m
armv7l
```

# Resizing an arm Virtual Machine
```sh
$ arm_now resize +10G
WARNING: Image format was not specified for 'arm_now/rootfs.ext2' and probing guessed raw.
         Automatically detecting the format is dangerous for raw images, write operations on block 0 will be restricted.
         Specify the 'raw' format explicitly to remove the restrictions.
Image resized.
```

# Debug the ls binary on mips
```sh
$ arm_now start mips32el
Welcome to arm_now
buildroot login: root
# gdb /bin/ls
(gdb) start
Temporary breakpoint 1, 0x00405434 in main ()
(gdb) x/i $pc
=> 0x405434 <main+12>:	li	a0,-1
```

## How to exit qemu

Press "Ctrl + ]" (Ctrl + altgr + ] on azerty).

# Wiki
All features and good use cases are documented in the wiki: https://github.com/nongiach/arm_now/wiki
- Exploit a heap use after free on PowerPC
- Run the game of life on a FPGA cpu (MicroBlaze)
- Share files
- Install new package
- ..

# Supported cpu

| CPU | images |
| --- | --- |
| arm | armv5-eabi, armv6-eabihf, armv7-eabihf |
| m68k | m68k-coldfire |
| microblaze | microblazeel, microblazebe |
| mips | mips32, mips32el, mips32r5el, mips32r6el |
| mips64 | mips64-n32, mips64el-n32 |
| xtensa | xtensa (a configurable processor) |
| nios2 | nios2 |
| powerpc | powerpc64-e5500, powerpc64-power8, powerpc64le-power8 |
| sh4 | sh-sh4 |
| x86-32 | x86-core2, x86-i686 |
| x86-64 | x86-64-core-i7 |
| aarch64 | aarch64 |



## Writeups

| Link | Credit |
| --- | --- |
| [MIPS binary exploitation challenge](http://www.aperikube.fr/docs/breizhctf_2018_mips/) | [Aperikube](https://twitter.com/AperiKube) |

Ping me for any new writeups.

## Project using arm_now

| Project | Credit |
| --- | --- |
| [Mandibule: linux elf injector for x86 x86_64 arm arm64](https://github.com/ixty/mandibule) | [ixty](https://twitter.com/_ixty_) |


----
By [@chaignc][] [#HexpressoTeam][hexpresso].


[hexpresso]:     https://hexpresso.github.io
[@chaignc]:    https://twitter.com/chaignc


Changelog.txt
--- Wed Feb 15 2017
First commit
arm only running on windows and linux host.
dhcpd using scapy.

--- Wed Feb 28 2018
Move from makefile scripting to python.
Now supporting arm, m68k, microblaze, mips, nios2, powerpc, sh4, x86

--- Sat Mar 17 2018
Add package manager.

--- Mon Mar 26 2018
Add support for aarch64, microblazebe, mips32r5el, mips32r6el, xtensa
Fix static IP to use the right interface on all vm.
Fix network interfaces
Detect the current kernel arch and if it doesn't match the asked arch, prompt "clean?"

--- Sun Apr 8 2018
Fix the use of format string to be compatible with python3.5

--- Thu Apr 12 2018
Add tcp/udp --redir

--- Mon Apr 23 2018
Add --offline and --resize option by @NoobieDog

--- Thu May 31 2018
Correct --resize
Fix bug with trying to sync big files
Migrate from clize to docopt
Refactor utils functions
Refactor filesystem, config, options functions
Disable audio to gain boot speed
Add colors
Add -add-qemu-options
Add show command

--- Thu May 31 2018
TODO:
Add a new features sections
	autostart
	resize
	show
	https
	full support of ext4
	Package manager is preinstalled

	we can share drive like this:
arm_now start x86-core2 --add-qemu-options "-machine usb=on -usb -usbdevice disk:empty-ext3-2GB.img"

find binary-samples/ -type f |tee filelist | e2cp -apv -d arm_now/rootfs.ext2:/
10.0.2.2 is the host ip
use mkisofs or genisoimage to create an iso from a folder. 
https://unix.stackexchange.com/questions/90793/create-iso-image-from-folder-via-terminal-commands
sudo mount -o loop output_image.iso iso/

Add a wiki page describing every options, --redir ..  and use of alias
Unit test
For --offline option check if local filesystem are avaiable otherwise exit cleanly and print a error message.
For --resize option check with a regex if the parameter is correct to avoid command/parameter injection.
Search for new kernel and rootfs images, maybe I missed some of them that have ramfs inside the kernel image.
real time shared folder using libvirt ?
enable kvm for x86
Add cpio filesystem full support (now --sync and host config is only supported
Refactor filesystem code.
A more flexible argument parsing.
Add bridged network.
rootfs are mostly composed of null bytes, so we can download, preconfig, and compress all of them to gain speed.
Use dhcp to connect several vm together
use etc/initab for automatic sync saving ?

--- Mon 24 Aug 2020
Fixing autostart with sync, so synced files can be used in the autostart script.
Adding unit test, just start => ./unit_test_all.sh



arm_now/config.py
class Config:
    DIR = "arm_now/"
    kernel = "kernel"
    dtb = "dtb"
    rootfs = "rootfs.ext2"
    arch = "arch"
    KERNEL = DIR + "kernel"
    DTB = DIR + "dtb"
    ROOTFS = DIR + "rootfs.ext2"
    ARCH = DIR + "arch"
    DOWNLOAD_CACHE_DIR = "/tmp/arm_now"

qemu_options = {
        "aarch64": ["aarch64", "-M virt -cpu cortex-a57 -smp 1 -kernel {kernel} -append 'root=/dev/vda console=ttyAMA0' -netdev user,id=eth0 -device virtio-net-device,netdev=eth0 -drive file={rootfs},if=none,format=raw,id=hd0 -device virtio-blk-device,drive=hd0"],
        "armv5-eabi": ["arm", "-M vexpress-a9 -kernel {kernel} -sd {rootfs} -append 'root=/dev/mmcblk0 console=ttyAMA0 rw physmap.enabled=0 noapic'"], # check log
        "armv6-eabihf": ["arm", "-M vexpress-a9 -kernel {kernel} -sd {rootfs} -append 'root=/dev/mmcblk0 console=ttyAMA0 rw physmap.enabled=0 noapic'"], # check log
        "armv7-eabihf": ["arm", "-M vexpress-a9 -kernel {kernel} -sd {rootfs} -append 'root=/dev/mmcblk0 console=ttyAMA0 rw physmap.enabled=0 noapic'"], # check log
        # "bfin":, TODO
        # "m68k-68xxx":, TODO 
        "m68k-coldfire": ["m68k", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/sda console=ttyS0 rw physmap.enabled=0 noapic'"], # check log
        "microblazebe": ["microblaze", "-M petalogix-s3adsp1800 -kernel {kernel} -nographic"], # rootfs is inside the kernel file, but we also have a separated rootfs if needed
        "microblazeel": ["microblazeel", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/sda console=tty0 rw physmap.enabled=0 noapic'"], # check log
        "mips32": ["mips", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/hda console=ttyS0 rw physmap.enabled=0 noapic'"], # check log
        "mips32el": ["mipsel", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/hda console=ttyS0 rw physmap.enabled=0 noapic'"], # check log
        "mips32r5el": ["mipsel", "-machine malta -cpu P5600 -kernel {kernel} -drive file={rootfs},format=raw -append 'root=/dev/hda rw'"],
        "mips32r6el": ["mipsel", "-M malta -cpu mips32r6-generic -kernel {kernel} -drive file={rootfs},format=raw -append root=/dev/hda -net nic,model=pcnet -net user"],
        "mips64-n32": ["mips64", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/hda console=ttyS0 rw physmap.enabled=0 noapic'"], # check log
        "mips64el-n32": ["mips64el", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/hda console=ttyS0 rw physmap.enabled=0 noapic'"], # check log
        # "mips64r6el-n32":, TODO check log
        # "mips64r6el-n32": ["mips64el", "-machine malta -kernel {kernel} -drive file={rootfs},format=raw -append 'root=/dev/hda rw console=ttyS0,'"], # check log
        "nios2": ["nios2", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/sda console=ttyS0 rw physmap.enabled=0 noapic'"], # check log
        "powerpc64-e5500": ["ppc64", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/sda console=ttyS0 rw physmap.enabled=0 noapic'"], # check log
        "powerpc64-power8": ["ppc64", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/sda console=ttyS0 rw physmap.enabled=0 noapic'"], # check log
        "powerpc64le-power8": ["ppc64", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/sda console=ttyS0 rw physmap.enabled=0 noapic'"], # check log
        "sh-sh4": ["sh4", "-M r2d -serial vc -kernel {kernel} -hda {rootfs} -append 'root=/dev/sda console=ttyS0 rw physmap.enabled=0 noapic'"], # check log
        # "sparc64":, TODO check log
        # "sparc64": ["sparc64", "-M sun4u -kernel {kernel} -append 'root=/dev/sda console=ttyS0,115200' -drive file={rootfs},format=raw -net nic,model=e1000 -net user"], # this causes kernel crash
        # ":sparcv8":, TODO, check log, 
        # "sparcv8": ["sparc", "-machine SS-10 -kernel {kernel} -drive file={rootfs},format=raw -append 'root=/dev/sda console=ttyS0,115200' -net nic,model=lance -net user"], # error
        # "x86-64-core-i7":["x86_64", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/sda console=ttyS0 rw physmap.enabled=0 noapic'"], # old
        "x86-64-core-i7" : ["x86_64", "-M pc -kernel {kernel} -drive file={rootfs},if=virtio,format=raw -append 'root=/dev/vda rw console=ttyS0' -net nic,model=virtio -net user"],
        # "x86-core2" : ["i386", "-kernel {kernel} -hda {rootfs} -append 'root=/dev/sda console=ttyS0 rw physmap.enabled=0 noapic -net nic,model=virtio -net user'"],
        "x86-core2": ["i386", "-M pc -kernel {kernel} -drive file={rootfs},if=virtio,format=raw -append 'root=/dev/vda rw console=ttyS0' -net nic,model=virtio -net user"], # fix opkg
        "x86-i686":["i386", "-M pc -kernel {kernel} -drive file={rootfs},if=virtio,format=raw -append 'root=/dev/vda rw console=ttyS0' -net nic,model=virtio -net user"],
        "xtensa-lx60": ["xtensa", "-M lx60 -cpu dc233c -monitor null -nographic -kernel {kernel} -monitor null"]
        }

""" The final user is not vulnerable to MITM attack, thoose http links are used to preconfigure the package manager.
When you do a arm_now start armv5-eabi the image you are downloading already has opkg installed and
everything is configured to use https, zero request are made over http.
thx lucasduffey
"""

install_opkg = {
        "armv5-eabi":"""wget -O - http://pkg.entware.net/binaries/armv5/installer/entware_install.sh | /bin/sh""",
        "armv7-eabihf":"""wget -O - http://pkg.entware.net/binaries/armv5/installer/entware_install.sh | /bin/sh""",
        "mips32el":"""wget -O - http://pkg.entware.net/binaries/mipsel/installer/installer.sh | /bin/sh""",
        "x86-64-core-i7":"""wget -O - http://pkg.entware.net/binaries/x86-64/installer/entware_install.sh | /bin/sh""",
        "x86-core2":"""wget -O - http://pkg.entware.net/binaries/x86-32/installer/entware_install.sh | /bin/sh""",
        "x86-i686":"""wget -O - http://pkg.entware.net/binaries/x86-32/installer/entware_install.sh | /bin/sh""",
}
